openstf_dbc/services/predictor.py

```diff
@@ -2,36 +2,137 @@
+
+        if PredictorGroups.WEATHER_DATA in predictor_groups and location is None:
+            raise ValueError(
+                "Need to provide a location when weather data predictors are requested."
+            )
+
+        predictors = pd.DataFrame(
+            index=get_datetime_index(datetime_start, datetime_end, forecast_resolution)
+        )
+
+        if PredictorGroups.WEATHER_DATA in predictor_groups:
+            weather_data_predictors = self.get_weather_data(
+                datetime_start,
+                datetime_end,
+                location=location,
+                forecast_resolution=forecast_resolution,
+            )
+            predictors = pd.concat([predictors, weather_data_predictors], axis=1)
+
+        if PredictorGroups.MARKET_DATA in predictor_groups:
+            market_data_predictors = self.get_market_data(
+                datetime_start, datetime_end, forecast_resolution=forecast_resolution
+            )
+            predictors = pd.concat([predictors, market_data_predictors], axis=1)
+
+        if PredictorGroups.LOAD_PROFILES in predictor_groups:
+            load_profiles_predictors = self.get_load_profiles(
+                datetime_start, datetime_end, forecast_resolution=forecast_resolution
+            )
+            predictors = pd.concat([predictors, load_profiles_predictors], axis=1)
+
+        return predictors
+
+    def get_market_data(self, datetime_start, datetime_end, forecast_resolution=None):
+        electricity_price = self.get_electricity_price(
+            datetime_start, datetime_end, forecast_resolution
+        )
+        gas_price = self.get_gas_price(
+            datetime_start, datetime_end, forecast_resolution
+        )
+
+        if electricity_price.empty is False and gas_price.empty is True:
+            return electricity_price
+
+        if electricity_price.empty is True and gas_price.empty is False:
+            return gas_price
+
+        if electricity_price.empty is True and gas_price.empty is True:
+            return pd.DataFrame(
+                index=get_datetime_index(
+                    datetime_start, datetime_end, forecast_resolution
+                )
+            )
+
+        return pd.concat([electricity_price, gas_price], axis=1)
+
+    def get_electricity_price(
+        self, datetime_start, datetime_end, forecast_resolution=None
+    ):
         query = 'SELECT "Price" FROM "forecast_latest".."marketprices" \
         WHERE "Name" = \'APX\' AND time >= \'{}\' AND time <= \'{}\''.format(
             datetime_start, datetime_end
         )
 
-        result = _DataInterface.get_instance().exec_influx_query(query)
+        electricity_price = _DataInterface.get_instance().exec_influx_query(query)
+
+        electricity_price = electricity_price["marketprices"]
+
+        electricity_price.rename(columns=dict(Price="APX"), inplace=True)
+
+        if forecast_resolution and electricity_price.empty is False:
```

Comment:
If the influx query yields no results, a hard-to-interpret exception is raised on line 111 (when selecting 'marketprices'), hence in the old code, this was after an `if result:` statement. Perhaps it would help us to still check if a result is returned explicitly before the column selection

Reply:
I guess this is related to my original question? Do we allow every predictor to be not available in the database. If so I should add some more checks.

Reply:
I.m.o. we allow every predictor to be not available, as I tried to describe in the general PR comment

Reply:
Sounds good. I've made a couple changes. All get predictor function now return an empty dataframe with a valid index if no data is returned from the database (influx query this is an empty dict, mysql query this is an empty dataframe).